### PR TITLE
add template for pull requests

### PR DESCRIPTION
**Pull request by mpapis**
- related to #1497
- based on https://github.com/blog/2111-issue-and-pull-request-templates

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Description:
+
+# Tasks:
+
+- [ ] Describe the problem / feature
+- [ ] Write tests for it
+- [ ] Write code to solve the problem
```
Comment (krainboltgreene) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'm not sure we need this one.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Description:
+
+# Tasks:
+
+- [ ] Describe the problem / feature
+- [ ] Write tests for it
```
Comment (krainboltgreene) [inline .github/PULL_REQUEST_TEMPLATE.md]: Drop `for it` IMO.

**Comment by krainboltgreene** [comment]
Last suggestion: Add a link to the COC like the issue template. I think it's great to have that at the top of every discussion.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (mpapis) [inline .github/PULL_REQUEST_TEMPLATE.md]: my intention is to show order, many developers jump in and fix a problem without writing a test to prove first there is a problem in first place - and I'm guilty of that one too, having explicit guide to write tests first might help.

**Comment by mpapis** [comment]
added placeholder for description, dropped "for it", and added COC

**Comment by mpapis** [comment]
I wonder where to put information about commits and including `[ci skip]` in case the change does not include testable code - like this readme like changes

**Comment by krainboltgreene** [comment]
Honestly, I'm OK with never skipping CI unless it's from us. @indirect or @segiddins?

**Comment by segiddins** [comment]
Nothing should EVER make it onto master without running on CI. Ever.

**Comment by krainboltgreene** [comment]
Agreed, though some people make exceptions for pure documentation @segiddins.

**Comment by segiddins** [comment]
@krainboltgreene I personally prefer no exceptions -- then you don't even have to think about it!

**Comment by krainboltgreene** [comment]
Yes, I definitely agree.

**Comment by mpapis** [comment]
ok, anything else I should update?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+# Tasks:
+
+- [ ] Describe the problem / feature
+- [ ] Write tests
+- [ ] Write code to solve the problem
+- [ ] Get code review from coworkers / friends
+- [ ] [Squash commits](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)
+
+I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).
```
Comment (simi) [inline .github/PULL_REQUEST_TEMPLATE.md]: Are you sure we really need this line?

Comment (mpapis) [inline .github/PULL_REQUEST_TEMPLATE.md]: yes, it's explicitly stating the rules on this project, now no one can say "I did not know"

Comment (simi) [inline .github/PULL_REQUEST_TEMPLATE.md]: Can you point me to some PR where this was needed?

Comment (mpapis) [inline .github/PULL_REQUEST_TEMPLATE.md]: No I can not, I'm not maintainer and it's not my place to know such situations, I was asked by one of the maintainers to add it - and I also believe it fits here good.

Please read https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md the line added here only ensures it's known to everybody before something that would invalidate their contribution.

**Comment by duckinator** [comment]
This looks good to me; :+1:. @krainboltgreene?

**Comment by krainboltgreene** [comment]
Yup. Have at it!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 3 above)
Comment (drbrain) [inline .github/PULL_REQUEST_TEMPLATE.md]: I didn't justify adding the CoC in the first place and we don't need to justify adding it to the issue template either.

**Comment by indirect** [comment]
One follow-up point: if we genuinely want squashed commits in PRs (do we? that's not written down anywhere else, and we don't seem to have enforced it), we can just check a box in the Homu settings for auto squashing.

**Comment by mpapis** [comment]
it was not required to squash in the past, but I see a lot of commits/comments that are pointless to the history and would be better squashed to the the commit that actually does the change, in some cases like fixing multiple separate issues in one PR like in beginning of https://github.com/rubygems/rubygems/pull/1355/commits - it makes sense to keep for good history overview - but look at the end - there are fixes for previous commits - which clutters history.

Maybe the text should say:
- squash fix up commits
- or: optionally squash commits
?

**Comment by drbrain** [comment]
We should remove "squash commits".

If the original author needed to separate the commits to keep their work straight how is someone else supposed to figure out what they were doing when fixing a bug in if we've destroyed the context of the commits?

**Comment by duckinator** [comment]
I agree with @drbrain on this. Not squashing commits makes the history more noisy, but it also makes the history more _accurate to how the development was done_, which (as @drbrain said) can make it easier to understand context when investigating bugs.

**Comment by mpapis** [comment]
I have to say that my intention in most cases was not the history but laziness, squashing commits could be a bit complicated tasks if you need to add changes to past commits, maybe we should just expand on what the squashing is supposed to do: grouping similar commits, maybe better tutorial with squashing two groups of commits? example: commits [1,2,3,4], squash 1+3, 2+4, WDYT?

**Comment by duckinator** [comment]
I feel like at that point it's just adding noise and complexity to the contributing process. We really want to streamline that so that we can accept, give feedback on, or reject things quickly.

Also: developer laziness is a valid thing to consider &mdash; are people going to bother? are maintainers going to consistently enforce it? If not, remove it from the task list to keep things simple.

**Comment by duckinator** [comment]
So this basically comes down to noise in the contributing process (squashing related consecutive commits) vs noise in history (no squashed commits).

The issue here is that the "middle ground" (squashing all commits) falls flat on both accounts, and as such is pretty much useless, so we have to actually choose one. I personally feel we want to make the contributing process as simple as possible for both contributors and the people reviewing the pull requests, but definitely see the case for squashed commits, so I'm okay with either. But if we do want squashed commits, we'll need to explain what we mean and properly document how to do it, as @mpapis said. (Frankly, I don't know how to do it properly.)

**Comment by mpapis** [comment]
first of all I'm fine if we remove it.

that said, we require adding tests, I think adding tests to rubygems is a lot more complicated then squashing, especially for somebody new ... on other hand it's actually adding extra effort that was not required earlier

**Comment by drbrain** [comment]
I've worked on my own code where I made a massive commit with a message that was insufficiently descriptive of why some important line changed.  It took quite some time to figure out why I had changed that one line and I think more-granular commits, even if they still didn't have insufficient descriptions, would have reduced the time spent regaining context.

I'm reasonably certain that someone else looking at my (mostly older) giant commits will come to feel the same way as I did.

**Comment by mpapis** [comment]
@drbrain I was thinking to use https://github.com/rubygems/rubygems/pull/1355/commits as example where I failed to squash some commits, but only `address comments` broke the rank, rest of the commits fits the purpose of describing what's going on

**Comment by simi** [comment]
That's hard to say. Sometimes it is useful to squash commits, sometimes it is not. I think it is ok to remove this line and ask for squash in PR when needed.

**Pull request merged**